File: solutions/csharp/allergies/1/Allergies.cs
using System;
using System.Collections.Generic;

[Flags]
public enum Allergen
{
    Eggs = 1,
    Peanuts = 2,
    Shellfish = 4,
    Strawberries = 8,
    Tomatoes = 16,
    Chocolate = 32,
    Pollen = 64,
    Cats = 128
}

public class Allergies
{
    private Allergen _mask;
    public Allergies(int mask)
    {
        _mask = (Allergen)mask;
    }

    public bool IsAllergicTo(Allergen allergen)
    {
        return _mask.HasFlag(allergen);
    }

    public Allergen[] List()
    {
        var list = new List<Allergen>();
        foreach (var item in Enum.GetNames(typeof(Allergen)))
        {
            var flag = (Allergen)Enum.Parse(typeof(Allergen), item);
            if (_mask.HasFlag(flag)) {
                list.Add(flag);
            }
        }
        return list.ToArray();
    }
}

File: solutions/fortran/allergies/1/allergies.f90

module allergies
  implicit none
  character(len=12) :: allergens(8) = (/'eggs        ', &
  'peanuts     ','shellfish   ','strawberries', &
  'tomatoes    ','chocolate   ','pollen      ','cats        '/)

contains

  logical function allergicTo(allergy_str, allergy_key)
    character(len=*), intent(in) :: allergy_str
    integer, intent(in) :: allergy_key
    character(len=64) :: binary
    integer :: i
    write(binary,'(b0)') allergy_key
    i =  len_trim(binary) + 1
    i = i - findloc(allergens, allergy_str, dim=1)
    allergicTo = binary(i:i) .eq. '1'
  end function


  function allergicList(allergy_key)
    integer, intent(in) :: allergy_key
    character(len=100) :: allergicList
    integer :: i
    allergicList = ' '
    do i = 1, 8
      if (allergicTo(allergens(i), allergy_key)) then
        allergicList = trim(allergicList) // ' ' // allergens(i)
      end if
    end do
    allergicList = adjustl(allergicList)
  end function



end module
